##### Z-test

In [4]:
import numpy as np
from scipy import stats

# Given values for the z-test
mean_new_drug = 10  # Mean improvement for new drug group
mean_placebo = 5    # Mean improvement for placebo group
std_new_drug = 5    # Standard deviation for new drug group
std_placebo = 5     # Standard deviation for placebo group
n_new_drug = 50     # Sample size for new drug group
n_placebo = 50      # Sample size for placebo group

# Calculate the z-score
z = (mean_new_drug - mean_placebo) / np.sqrt(
    (std_new_drug**2 / n_new_drug) + (std_placebo**2 / n_placebo))

# Calculate the p-value (two-tailed test)
p_value = 2 * (1 - stats.norm.cdf(abs(z)))

# Output the results
print(f"Z-score: {z}")
print(f"P-value: {p_value}")

# Decision based on a significance level (alpha)
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis (there is a significant difference between the groups).")
else:
    print("Fail to reject the null hypothesis (no significant difference between the groups).")

Z-score: 5.0
P-value: 5.733031438470704e-07
Reject the null hypothesis (there is a significant difference between the groups).


In [5]:
# sample Data - Average population weight in kgs
data = [78, 62, 80, 75, 61, 74, 79, 93, 77, 95, 67, 88, 54, 69]

In [6]:
# Define the null hypothesis (average population weight = 80)
null_mean = 80

In [7]:
len(data)

14

In [8]:
# Calculate the sample mean and standard deviation
sample_mean = np.mean(data)
sample_std = np.std(data)

In [9]:
sample_mean

75.14285714285714

In [10]:
sample_std

11.506874608915581

In [11]:
# Calculate the z-score
z_score = (sample_mean - null_mean) / sample_std

In [12]:
z_score

-0.4221079156784696

In [13]:
# Calculate the two-tailed p-value using the z-score
p_value = 2 * (1 - stats.norm.cdf(abs(z_score)))

In [14]:
p_value

0.6729462504556214

In [15]:
# Define significance level (alpha) - commonly used value is 0.05 (5%)
alpha = 0.05

In [16]:
# Interpretation
if p_value < alpha:
  print('''Reject the null hypothesis. The average weight is statistically different 
  from 80 at a significance level of''', alpha)
else:
  print('''Fail to reject the null hypothesis. There is not enough evidence to conclude 
  the average weight is statistically different from 80 at a significance level of''', alpha)

Fail to reject the null hypothesis. There is not enough evidence to conclude 
  the average weight is statistically different from 80 at a significance level of 0.05


In [18]:
from statsmodels.stats.weightstats import ztest

In [19]:
#enter IQ levels for 20 patients
data = [88, 92, 94, 94, 96, 97, 97, 97, 99, 99,
        105, 109, 109, 109, 110, 112, 112, 113, 114, 115]

In [20]:
#perform one sample z-test
ztest(data, value=100)

(1.5976240527147705, 0.1101266701438426)

The test statistic for the one sample z-test is 1.5976 and the corresponding p-value is 0.1101.

Since this p-value is not less than .05, we do not have sufficient evidence to reject the null hypothesis. In other words, the new drug does not significantly affect IQ level.

In [21]:
pip install statsmodels

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.1.1 -> 24.2
[notice] To update, run: python3 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


##### 𝛘-Square Test

In [26]:
from scipy.stats import chi2_contingency
import numpy as np

In [29]:
# defining the table
data = np.array([[207, 282, 241], [234, 242, 232]])

In [30]:
data

array([[207, 282, 241],
       [234, 242, 232]])

In [31]:
data.sum(axis=0)

array([441, 524, 473])

In [32]:
data.sum(axis=1)

array([730, 708])

In [33]:
stat, p, dof, expected = chi2_contingency(data)

In [34]:
stat

4.542228269825232

In [35]:
p

0.1031971404730939

In [36]:
dof

2

In [37]:
expected

array([[223.87343533, 266.00834492, 240.11821975],
       [217.12656467, 257.99165508, 232.88178025]])

In [39]:
# interpret p-value
alpha = 0.05
print("p value is " + str(p))
if p <= alpha:
    print('Dependent (reject H0)')
else:
    print('Independent (H0 holds true)')

p value is 0.1031971404730939
Independent (H0 holds true)


In [40]:
#### Example

In [41]:
# defining the table
data = np.array([[5, 10, 5], [15, 5, 0],[5, 10, 15]])
stat, p, dof, expected = chi2_contingency(data)

In [43]:
p

5.540229051644113e-05

In [42]:
# interpret p-value
alpha = 0.05
print("p value is " + str(p))
if p <= alpha:
    print('Dependent (reject H0)')
else:
    print('Independent (H0 holds true)')

p value is 5.540229051644113e-05
Dependent (reject H0)
